feat(button): add loadingPosition prop for loading icon placement control

yeonjulee1005 opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue
#5362 
<!-- If it resolves an open issue, please link the issue here. For example "Resolves #123" -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Adds a new `loadingPosition` prop to the Button component that allows developers to control where the loading icon appears when the button is in a loading state.

**Features:**
- Three position options: `left`, `center`, and `right`
- `left`: displays loading icon on the left (default behavior)
- `center`: displays loading icon in the center, hiding the label
- `right`: displays loading icon on the right


https://github.com/user-attachments/assets/818c22ed-09dd-4a54-87ff-085fc5ce4b09


**Backward Compatibility:**
- Maintains full backward compatibility with existing code
- Existing usage with `leading` and `trailing` props continues to work without any changes
- When `loadingPosition` is not specified, the component automatically determines the position based on the `trailing` prop

**Implementation Details:**
- Added `loadingPosition` prop to `ButtonProps` interface
- Updated `useComponentIcons` logic to handle position-based loading icon placement
- Added conditional rendering for center position (hides label, shows only loading icon)
- Added comprehensive test cases for all three position options

**Documentation:**
- Added new "Loading Position" section in button.md
- Includes code examples for each position option
- Added backward compatibility note

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

yeonjulee1005 committed:
feat(button): add loadingPosition prop to control loading icon placement

Add a new `loadingPosition` prop to the Button component that allows
developers to control where the loading icon appears when the button is
in a loading state. The prop accepts three values:
- 'left': displays loading icon on the left (default behavior)
- 'center': displays loading icon in the center, hiding the label
- 'right': displays loading icon on the right

The implementation maintains full backward compatibility. Existing usage
with `leading` and `trailing` props for loading buttons continues to
work without any changes. When `loadingPosition` is not specified, the
component automatically determines the position based on the `trailing`
prop, preserving the existing behavior.

Adds test cases for all three loadingPosition values to ensure proper
rendering and behavior.

yeonjulee1005 committed:
docs(button): add loadingPosition prop documentation

Add documentation section for the new `loading-position` prop in the
Button component. The section includes:
- Description of the three position options (left, center, right)
- Code examples for each position option
- Backward compatibility note explaining that existing usage with
  `leading` and `trailing` props continues to work

This documentation helps developers understand how to control the
loading icon placement when using the Button component in loading state.

yeonjulee1005 committed:
chore(button): remove comments

yeonjulee1005 committed:
chore: add .vscode to .gitignore and remove from tracking

yeonjulee1005 committed:
test(button): update Vue snapshots for loadingPosition

yeonjulee1005 commented:
Thank you! :)

rijenkii commented:
First off -- thank you. I have some comments and questions.

1. I'm not sure whether this needs to be a `UButton`-specific feature, or if it could be useful for other components and should instead be implemented in `useComponentIcons` -- when I was making the issue I have not read icon positioning source code, but now that I have I see that the icon positioning logic is reused between `Badge`, `Input*`, `Select*` and `Textarea`, and it may be useful there as well.

2. In the documentation you say:
   > Use the `loading-position` prop to control where the loading icon appears. The prop accepts three values:
   > - `left`: displays loading icon on the left (default)

   In but in the issue you say:
   > **Backward Compatibility:**
   > - When loadingPosition is not specified, the component automatically determines the position based on the trailing prop

   Which one is it? Current behaviour is not putting the loading spinner at the start of the button, but instead follows a complex logic that I was not able to simplify much yet:
   <details><summary>Screenshot</summary><img width="1904" height="1304" alt="Screenshot 2025-11-15 at 12-55-15 " src="https://github.com/user-attachments/assets/596cbc93-7d4c-4e94-a8c8-fafd54d902b8" /></details>

   <details><summary>Source</summary>

   ```vue
   <script setup lang="ts">
   const args: {
     icon?: string;
     leading?: boolean;
     trailing?: boolean;
     leadingIcon?: string;
     trailingIcon?: string;
   }[] = [
     {},
     { trailingIcon: "i-lucide-chevron-right" },
     { trailing: true },
     { trailing: true, trailingIcon: "i-lucide-chevron-right" },
     { leadingIcon: "i-lucide-chevron-left" },
     {
       leadingIcon: "i-lucide-chevron-left",
       trailingIcon: "i-lucide-chevron-right",
     },
     { leadingIcon: "i-lucide-chevron-left", trailing: true },
     {
       leadingIcon: "i-lucide-chevron-left",
       trailing: true,
       trailingIcon: "i-lucide-chevron-right",
     },
     { leading: true },
     { leading: true, trailingIcon: "i-lucide-chevron-right" },
     { leading: true, trailing: true },
     { leading: true, trailing: true, trailingIcon: "i-lucide-chevron-right" },
     { leading: true, leadingIcon: "i-lucide-chevron-left" },
     {
       leading: true,
       leadingIcon: "i-lucide-chevron-left",
       trailingIcon: "i-lucide-chevron-right",
     },
     { leading: true, leadingIcon: "i-lucide-chevron-left", trailing: true },
     {
       leading: true,
       leadingIcon: "i-lucide-chevron-left",
       trailing: true,
       trailingIcon: "i-lucide-chevron-right",
     },
     { icon: "i-lucide-chevrons-left-right" },
     {
       icon: "i-lucide-chevrons-left-right",
       trailingIcon: "i-lucide-chevron-right",
     },
     { icon: "i-lucide-chevrons-left-right", trailing: true },
     {
       icon: "i-lucide-chevrons-left-right",
       trailing: true,
       trailingIcon: "i-lucide-chevron-right",
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leadingIcon: "i-lucide-chevron-left",
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leadingIcon: "i-lucide-chevron-left",
       trailingIcon: "i-lucide-chevron-right",
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leadingIcon: "i-lucide-chevron-left",
       trailing: true,
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leadingIcon: "i-lucide-chevron-left",
       trailing: true,
       trailingIcon: "i-lucide-chevron-right",
     },
     { icon: "i-lucide-chevrons-left-right", leading: true },
     {
       icon: "i-lucide-chevrons-left-right",
       leading: true,
       trailingIcon: "i-lucide-chevron-right",
     },
     { icon: "i-lucide-chevrons-left-right", leading: true, trailing: true },
     {
       icon: "i-lucide-chevrons-left-right",
       leading: true,
       trailing: true,
       trailingIcon: "i-lucide-chevron-right",
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leading: true,
       leadingIcon: "i-lucide-chevron-left",
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leading: true,
       leadingIcon: "i-lucide-chevron-left",
       trailingIcon: "i-lucide-chevron-right",
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leading: true,
       leadingIcon: "i-lucide-chevron-left",
       trailing: true,
     },
     {
       icon: "i-lucide-chevrons-left-right",
       leading: true,
       leadingIcon: "i-lucide-chevron-left",
       trailing: true,
       trailingIcon: "i-lucide-chevron-right",
     },
   ];
   </script>
   
   <template>
     <div class="flex flex-col gap-2 p-4">
       <div v-for="(arg, key) of args" :key class="flex gap-2">
         <UButton
           label="Label"
           :icon="arg.icon"
           :leading="arg.leading"
           :trailing="arg.trailing"
           :leading-icon="arg.leadingIcon"
           :trailing-icon="arg.trailingIcon"
         />
         <UButton
           label="Label"
           loading
           :icon="arg.icon"
           :leading="arg.leading"
           :trailing="arg.trailing"
           :leading-icon="arg.leadingIcon"
           :trailing-icon="arg.trailingIcon"
         />
         <code>{{ arg }}</code>
       </div>
     </div>
   </template>
   ```

   </details>

4. No sure whether `left`/`right` is better than `leading`/`trailing` -- firstly because it introduces new language when `leading`/`trailing` is already there, secondly because of potential issues with RTL locales, but not sure about that one.

yeonjulee1005 commented:
@rijenkii 

Questions are always welcome!🤗

1. That’s right! There are other components that also receive `loading` prop.
However, since these components don’t handle events like buttons do, I believe the provided `value should always be displayed by default`.

2. I wanted to make sure existing users don’t get confused.
So while keeping the current logic where leading and trailing are chosen when using loading, I thought adding `loadingPosition` should still allow users to specify the intended placement.

3. Other components also use left and right as prop values, so I felt that using those(include `center` also) would be clearer and more consistent than using leading or trailing.

rijenkii commented:
I just looked at the implementation, and `loading-position="center"` does not do what I described in https://github.com/nuxt/ui/issues/5362: it does not keep the width of the button intact when toggling the loading.

The implemented behaviour (label disappears completely and is replaced by the loading icon, changing the width of the button):

https://github.com/user-attachments/assets/d75581a3-2c64-4698-9fc4-2511664e323f

The behaviour described in the issue (label and other icons receive `opacity: 0` with the spinner rendered above them, keeping the width of the button):

https://github.com/user-attachments/assets/e21e8eeb-ffd9-4260-9737-d166d9890eb3

yeonjulee1005 commented:
@rijenkii 
Sure~! I’ll update it so that the changes are applied.

https://github.com/user-attachments/assets/21494528-e4a5-41cf-99dc-5bb47feace01

yeonjulee1005 committed:
fix(button): maintain label width when loadingPosition is center

When loadingPosition is set to 'center' and a label is provided, the button now maintains the label's width even though the label is hidden. The loading icon is displayed in the center of the button while preserving the original button width.

Changes:
- Add loadingIcon and loadingLabel slots to button theme
- Add animate-spin class to loadingIcon when loading is true
- Update Button component to render label with opacity-0 and h-0 when loadingPosition is center
- Update test snapshots to reflect the new structure

yeonjulee1005 committed:
Merge branch 'v4' into feat(button)/add-loading-position-center

MuhammadM1998 commented:
I think it's better to rename `start` and `end` to `leading` and `trailing`, so that it makes sense in RTL as well. We're also using leading and trailing in icons props too, so it aligns better with the existing namings

yeonjulee1005 commented:
@MuhammadM1998 So what you’re saying is that rather than `start` and `end`, we should use `leading` and `trailing` to represent `left` and `right`, right?

I did set using `left` and `right` 😙

The reason I didn’t use `leading` and `trailing` is because those names are already used as `icon slot` names in `UButton`. 

I was concerned that this might confuse users, so I went with `left, center, and right` to clearly indicate the exact positions.

MuhammadM1998 commented:
Sorry my comment was confusing. You're not using `start` and `end` to begin with 😅. I meant renaming `left` and `right` to `leading` and `trailing`. 

`leading` and `trailing` are values for the prop and not the names, so it won't conflict with slots and it actually makes more sense as it aligns with the other conventions imo

My point is, positions are reversed in RTL. So if I set `loading-position` to `left` in LTR, I'd expect it to be in the **right** in RTL, hence the confusion that using `leading` and `trailing` would solve

yeonjulee1005 commented:
@MuhammadM1998 Got it now! You were highlighting the potential confusion for RTL/LTR users.

I’ll update it accordingly. Thanks for the suggestion!

yeonjulee1005 commented:
@MuhammadM1998 
It should be reflected as shown in the video below!

https://github.com/user-attachments/assets/375a016c-b84f-4e94-b0c9-7f2dfb1a7e62

yeonjulee1005 committed:
feat(button): change loadingPosition from left/right to leading/trailing with RTL support

Change the loadingPosition prop values from 'left'/'right' to 'leading'/'trailing' to better support RTL (Right-to-Left) layouts. The component now automatically reverses the loading icon position in RTL mode, ensuring proper layout regardless of text direction.

Changes:
- Update loadingPosition prop type from 'left' | 'center' | 'right' to 'leading' | 'center' | 'trailing'
- Update default value from 'left' to 'leading'
- Update internal logic to use leading/trailing instead of left/right
- Add RTL support to button theme with rtl:flex-row-reverse class
- Update block variant trailingIcon margin for RTL compatibility
- Update documentation to explain LTR/RTL behavior
- Update playground examples to use new prop values
- Add RTL/LTR toggle to button playground page for testing

yeonjulee1005 committed:
feat(button): change loadingPosition from left/right to leading/trailing with RTL support

Change the loadingPosition prop values from 'left'/'right' to 'leading'/'trailing' to better support RTL (Right-to-Left) layouts. The component now automatically reverses the loading icon position in RTL mode, ensuring proper layout regardless of text direction.

Changes:
- Update loadingPosition prop type from 'left' | 'center' | 'right' to 'leading' | 'center' | 'trailing'
- Update default value from 'left' to 'leading'
- Update internal logic to use leading/trailing instead of left/right
- Add RTL support to button theme with rtl:flex-row-reverse class on base slot
- Update block variant trailingIcon margin for RTL compatibility (ms-auto rtl:me-auto rtl:ms-0)
- Update documentation to explain LTR/RTL behavior for leading/trailing positions
- Update playground examples to use new prop values (leading/trailing)
- Update test files to use new prop values
- Remove obsolete left/right loadingPosition snapshots
- Add RTL/LTR toggle to button playground page for testing

MuhammadM1998 commented:
It looks great! thanks for the work on this 🙏

yeonjulee1005 commented:
@MuhammadM1998 Thanks!! ☺️

yeonjulee1005 committed:
Merge branch 'v4' into feat(button)/add-loading-position-center

yeonjulee1005 committed:
Merge branch 'v4' into feat(button)/add-loading-position-center

yeonjulee1005 committed:
fix(button): update snapshots to reflect Button RTL support

Changes:
- Update Slideover snapshots with rtl:flex-row-reverse class on close button elements
- Reflect Button component's RTL support in Slideover test snapshots
Changes:
- Add loadingIcon and loadingLabel slots to button theme
- Add animate-spin class to loadingIcon when loading is true
- Update Button component to render label with opacity-0 and h-0 when loadingPosition is center
- Update test snapshots to reflect the new structure

yeonjulee1005 committed:
Merge branch 'v4' into feat(button)/add-loading-position-center

yeonjulee1005 committed:
test(FormField): improve type safety and code readability

- Replace `as any` type assertions with explicit types (Size, Orientation)
- Extract test cases array to `testCases` variable for better readability
- Add explicit type annotations for map functions

benjamincanac committed:
chore(deps): update tiptap to v3.14.0

benjamincanac committed:
chore(deps): use tiptap versions from root package

benjamincanac committed:
chore(deps): use tiptap in docs and playground

benjamincanac committed:
chore(deps): remove `pnpm.overrides`

yeonjulee1005 committed:
Merge branch 'v4' into feat(button)/add-loading-position-center

yeonjulee1005 committed:
Merge branch 'v4' into feat(button)/add-loading-position-center

rijenkii commented:
Setting `loading-position="center"` completely removes all button icons:

<img width="1904" height="1304" alt="image" src="https://github.com/user-attachments/assets/809c50cd-6fc4-4ac5-bff9-b9b2aca7526d" />

<details><summary>Code</summary>

```vue
<script setup lang="ts">
const args: {
  icon?: string;
  leading?: boolean;
  trailing?: boolean;
  leadingIcon?: string;
  trailingIcon?: string;
}[] = [
  {},
  { trailingIcon: "i-lucide-chevron-right" },
  { trailing: true },
  { trailing: true, trailingIcon: "i-lucide-chevron-right" },
  { leadingIcon: "i-lucide-chevron-left" },
  {
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  { leadingIcon: "i-lucide-chevron-left", trailing: true },
  {
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { leading: true },
  { leading: true, trailingIcon: "i-lucide-chevron-right" },
  { leading: true, trailing: true },
  { leading: true, trailing: true, trailingIcon: "i-lucide-chevron-right" },
  { leading: true, leadingIcon: "i-lucide-chevron-left" },
  {
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  { leading: true, leadingIcon: "i-lucide-chevron-left", trailing: true },
  {
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right" },
  {
    icon: "i-lucide-chevrons-left-right",
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", trailing: true },
  {
    icon: "i-lucide-chevrons-left-right",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", leading: true },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", leading: true, trailing: true },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
];
</script>

<template>
  <div class="flex flex-col gap-2 p-4">
    <div v-for="(arg, key) of args" :key class="flex gap-2">
      <UButton
        label="Label"
        :icon="arg.icon"
        :leading="arg.leading"
        :trailing="arg.trailing"
        :leading-icon="arg.leadingIcon"
        :trailing-icon="arg.trailingIcon"
        loading-position="center"
      />
      <UButton
        label="Label"
        loading
        :icon="arg.icon"
        :leading="arg.leading"
        :trailing="arg.trailing"
        :leading-icon="arg.leadingIcon"
        :trailing-icon="arg.trailingIcon"
        loading-position="center"
      />
      <code>{{ arg }}</code>
    </div>
  </div>
</template>
```

</details>

Setting `loading-position="trailing"` does not work at all, behaviour is equivalent to not setting the parameter at all:

<img width="1904" height="1304" alt="image" src="https://github.com/user-attachments/assets/9f46801e-cb22-46e4-8e8a-206553199db7" />

<details><summary>Code</summary>

```vue
<script setup lang="ts">
const args: {
  icon?: string;
  leading?: boolean;
  trailing?: boolean;
  leadingIcon?: string;
  trailingIcon?: string;
}[] = [
  {},
  { trailingIcon: "i-lucide-chevron-right" },
  { trailing: true },
  { trailing: true, trailingIcon: "i-lucide-chevron-right" },
  { leadingIcon: "i-lucide-chevron-left" },
  {
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  { leadingIcon: "i-lucide-chevron-left", trailing: true },
  {
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { leading: true },
  { leading: true, trailingIcon: "i-lucide-chevron-right" },
  { leading: true, trailing: true },
  { leading: true, trailing: true, trailingIcon: "i-lucide-chevron-right" },
  { leading: true, leadingIcon: "i-lucide-chevron-left" },
  {
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  { leading: true, leadingIcon: "i-lucide-chevron-left", trailing: true },
  {
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right" },
  {
    icon: "i-lucide-chevrons-left-right",
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", trailing: true },
  {
    icon: "i-lucide-chevrons-left-right",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", leading: true },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", leading: true, trailing: true },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
];
</script>

<template>
  <div class="flex flex-col gap-2 p-4">
    <div v-for="(arg, key) of args" :key class="flex gap-2">
      <UButton
        label="Label"
        :icon="arg.icon"
        :leading="arg.leading"
        :trailing="arg.trailing"
        :leading-icon="arg.leadingIcon"
        :trailing-icon="arg.trailingIcon"
        loading-position="trailing"
      />
      <UButton
        label="Label"
        loading
        :icon="arg.icon"
        :leading="arg.leading"
        :trailing="arg.trailing"
        :leading-icon="arg.leadingIcon"
        :trailing-icon="arg.trailingIcon"
        loading-position="trailing"
      />
      <code>{{ arg }}</code>
    </div>
  </div>
</template>
```

</details>

Setting `loading-position="leading"` removes trailing icon when `:loading="true"`, if the trailing icon is set using a combination of `icon="..."` and `:trailing="true"`:

<img width="1904" height="1304" alt="image" src="https://github.com/user-attachments/assets/cc6163ea-42d0-4c45-892f-938c99688501" />

<details><summary>Code</summary>

```vue
<script setup lang="ts">
const args: {
  icon?: string;
  leading?: boolean;
  trailing?: boolean;
  leadingIcon?: string;
  trailingIcon?: string;
}[] = [
  {},
  { trailingIcon: "i-lucide-chevron-right" },
  { trailing: true },
  { trailing: true, trailingIcon: "i-lucide-chevron-right" },
  { leadingIcon: "i-lucide-chevron-left" },
  {
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  { leadingIcon: "i-lucide-chevron-left", trailing: true },
  {
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { leading: true },
  { leading: true, trailingIcon: "i-lucide-chevron-right" },
  { leading: true, trailing: true },
  { leading: true, trailing: true, trailingIcon: "i-lucide-chevron-right" },
  { leading: true, leadingIcon: "i-lucide-chevron-left" },
  {
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  { leading: true, leadingIcon: "i-lucide-chevron-left", trailing: true },
  {
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right" },
  {
    icon: "i-lucide-chevrons-left-right",
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", trailing: true },
  {
    icon: "i-lucide-chevrons-left-right",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", leading: true },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  { icon: "i-lucide-chevrons-left-right", leading: true, trailing: true },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailingIcon: "i-lucide-chevron-right",
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
  },
  {
    icon: "i-lucide-chevrons-left-right",
    leading: true,
    leadingIcon: "i-lucide-chevron-left",
    trailing: true,
    trailingIcon: "i-lucide-chevron-right",
  },
];
</script>

<template>
  <div class="flex flex-col gap-2 p-4">
    <div v-for="(arg, key) of args" :key class="flex gap-2">
      <UButton
        label="Label"
        :icon="arg.icon"
        :leading="arg.leading"
        :trailing="arg.trailing"
        :leading-icon="arg.leadingIcon"
        :trailing-icon="arg.trailingIcon"
        loading-position="leading"
      />
      <UButton
        label="Label"
        loading
        :icon="arg.icon"
        :leading="arg.leading"
        :trailing="arg.trailing"
        :leading-icon="arg.leadingIcon"
        :trailing-icon="arg.trailingIcon"
        loading-position="leading"
      />
      <code>{{ arg }}</code>
    </div>
  </div>
</template>
```

</details>

yeonjulee1005 commented:
@rijenkii 
Wow, thank you so much for the thorough testing!

Regarding my original intention:
- I intended to hide all icons when `loading-position="center"`.
- For leading or trailing positions, I thought the loading spinner should appear at the specified position (respecting the loading-position prop).

However, it seems `loading-position="trailing"` is definitely not working as intended.
Is my understanding correct?
I'll look into this and fix it ASAP!🥹

rijenkii commented:
> I intended to hide all icons when loading-position="center".

Then this PR is not implementing the functionality requested in https://github.com/nuxt/ui/issues/5362, see video in the linked issue -- it shows an icon on the button before it is switched into the loading state.

> For leading or trailing positions, I thought the loading spinner should appear at the specified position (respecting the loading-position prop).

Agreed.

> However, it seems `loading-position="trailing"` is definitely not working as intended.

Yes, and `loading-position="leading"` is also breaking in some cases, specifically these:

<img width="1904" height="184" alt="image" src="https://github.com/user-attachments/assets/95ff74d3-01b1-4dc0-a49c-2e72c610901b" />

In those cases trailing icon disappears when `loading` is enabled (left buttons are with `:loading="false"`, right are with `:loading="true"`).

yeonjulee1005 committed:
fix(Button): fix loading icon position and animation bugs

- Fix icons not displaying when loading-position="center" and loading=false
- Fix loading icon appearing in wrong position when loading-position="trailing"
- Prevent loading animation on regular icons when loading-position is set
- Maintain backward compatibility when loading-position is not specified

yeonjulee1005 committed:
Merge branch 'v4' into feat(button)/add-loading-position-center

yeonjulee1005 committed:
test(Button): update snapshots for loading icon position fixes

Snapshots updated to reflect correct loading icon positioning behavior
when loading-position prop is used.

yeonjulee1005 commented:
@rijenkii 

I’ve made the requested changes.
Please check the video for more details.
Thank you!🥹

https://github.com/user-attachments/assets/85b5ceb1-ce21-463c-8c67-c4b4ed62b295

yeonjulee1005 committed:
test(Button): update snapshots for loading icon position fixes

Snapshots updated to reflect correct loading icon positioning behavior
when loading-position prop is used.

rijenkii commented:
All `leading`, all `trailing` and `center` with `:loading="false"` work as I expected now. The only incorrect behaviour is `center` with `:loading="true"`.

Currently they look like this:
<img width="1904" height="184" alt="image" src="https://github.com/user-attachments/assets/5e0a2666-a87a-43c1-90d8-4731d0fbea5d" />

When I expected something like this:
<img width="1904" height="184" alt="image" src="https://github.com/user-attachments/assets/fe31c438-533d-4c8b-b45f-dd909bafe41f" />

Reason: the whole point of `loading-position="center"` is that the button should not change its size at all when `:loading` is toggled. Currently button is shrinking when `:loading` is set to `true` if the button has any icons.

benjamincanac commented:
Sorry I just got my eyes on this, your PR brings way too much complexity I will not be merging this. You can easily achieve this behavior using the Button's default slot!

rijenkii commented:
@benjamincanac will you not be willing to merge this functionality at all, or just this specific implementation? I have a desire to take a crack at this myself a bit later.

yeonjulee1005 commented:
@benjamincanac 

I fully understand that there are many snapshot changes because UButton is used in various places!
The reason I followed up was because there have been continuous branch updates for snapshots that don't depend on UButton.
Actually, I'm also aware that this feature can be implemented by clearly communicating the loading state through slots!
Thank you for reviewing this! 🥹

Regrettably, I've decided to close this PR!

benjamincanac commented:
@yeonjulee1005 It wasn't about the snapshots but more about the implementation in the Button, I understand lots of logic is necessary to achieve this but I have to make decisions to keep a good balance between functionality and maintainability. Sorry about that, hope you understand!

- https://github.com/nuxt/ui/pull/5456/changes#diff-14b97ad7ca8f3574aba9f092709d1305b0f214d25e7cc95062c8a290fa77bdf7R91-R195
- https://github.com/nuxt/ui/pull/5456/changes#diff-14b97ad7ca8f3574aba9f092709d1305b0f214d25e7cc95062c8a290fa77bdf7R218-R241
